internal/expr/bindtypes.go

```diff
@@ -44,80 +43,62 @@ func (pe *ParsedExpr) BindTypes(args ...any) (te *TypedExpr, err error) {
 		return nil, err
 	}
 
-	var sql bytes.Buffer
-
-	var inCount int
-	var outCount int
-
-	var outputs = make([]typeinfo.Output, 0)
-	var inputs = make([]typeinfo.Input, 0)
-
-	var outputUsed = make(map[typeinfo.Output]bool)
-
-	// Check and expand each query expr.
+	// Bind types to each expression.
+	typedExprs := []typedExpression{}
+	outputUsed := map[typeinfo.Output]bool{}
+	var te typedExpression
 	for _, expr := range pe.exprs {
 		switch e := expr.(type) {
 		case *inputExpr:
-			inLoc, err := bindInputTypes(argInfo, e)
+			te, err = bindInputTypes(e, argInfo)
```

Comment:
Here we go.

- Make `typedExpression` generic.
- It has a method `getTypedExpr[T]` or some such.
- So for example, `inputExpr` implements `typedExpression[typedInputExpr]`.
- `bindInputTypes` and `bindOutputTypes` become the bodies of `getTypedExpr`.
- `bypass` just returns itself.

Then it's starting to look nice.

Reply:
I would like to make the `bindInputTypes` and `bindOutputTypes` functions part of an interface, but I can't find a way to make the generics implementation work nicely.

If `expression` implements this generic interface then there is the problem that I cannot have a list of `typedExpression[T]` where `T` has a different instantiation at each index. This means I need to assert at some point that the expressions do actually implement this interface, and because they need to implement the one with the right instantiation of `T` I would still need to enumerate the types in a switch case statement.

I was thinking I could make `bindTypes` a method of expression, and it can return any so that we can get rid of the `typedExpression` interface.

Reply:
I've implemented `bindTypes` as a method on `expression`